Spring Security. Created a new endpoint for creating a user. Class BC…

StopSmile opened the pull request:
Class BCryptPasswordEncoder is used for encoding user passwords.

Commit:
Spring Security. Created a new endpoint for creating a user. Class BCryptPasswordEncoder is used for encoding user passwords.

Commit:
1.Add Brute Force protector. Block user email for 5 minutes on 3 unsuccessful logins.
2.Create an endpoint to show blocked users.

Des135 commented on Messenger/src/test/java/com/epam/testmodule/ChoseModeTest.java:
```diff
@@ -41,7 +41,7 @@ public void CheckInput_whenInputEqualsTwo_ShouldRunFileModerRun() throws IOExcep
         ChoseMode choseMode = new ChoseMode(consoleMode,fileMode);
         doReturn("C:\\Users\\Ivan_Pylypiv\\Desktop\\Library\\Messenger\\data2.txt").when(fileMode).getInfoFromUser();
-        doReturn(true).when(fileMode).checkFilePath("C:\\Users\\Ivan_Pylypiv\\Desktop\\Library\\Messenger\\data2.txt");
+        doReturn(true).when(fileMode).checkFilePath("C:\\Users\\Ivan_Pylypiv\\IdeaProjects\\Library\\Messenger\\data2.txt");
```
pls avoid using absolute file paths in the source code, as this project may be running on a different machine

Des135 commented on Messenger/src/test/java/com/epam/testmodule/FileModeTest.java:
```diff
@@ -42,10 +42,10 @@ void checkInput_WhenZeroIsProvided_ApplicationShouldExit() {
-        String file3 = "C:\\Users\\Ivan_Pylypiv\\Desktop\\Library\\Messenger\\data3.txt";
-        String file4 = "C:\\Users\\Ivan_Pylypiv\\Desktop\\Library\\Messenger\\data4.txt";
-        String file5 = "C:\\Users\\Ivan_Pylypiv\\Desktop\\Library\\Messenger\\data5.txt";
+        String file2 = "C:\\Users\\Ivan_Pylypiv\\IdeaProjects\\Library\\Messenger\\data2.txt";
```
same here, it would be better to use relative paths to resource files instead of absolute ones

Des135 commented on book-service-rest/src/main/java/com/example/Library/controller/UsersController.java:
```diff
@@ -0,0 +1,106 @@
+        this.userService = userService;
+        this.userRepository = userRepository;
+    }
+   // @PreAuthorize("hasAuthority('books:create')")
```
redundant comment could be removed (please keep your code clean)

Des135 commented on book-service-rest/src/main/java/com/example/Library/controller/UsersController.java:
```diff
@@ -0,0 +1,106 @@
+        userDTO.setAge(age);
+        userDTO.setEmail(email);
+        userDTO.setPassword(passwordEncoder.encode(password));
+        if (gender.equals("MALE")) {
```
It might be easier to have an enum or Set<String> with allowed values, so you can check if the input value is allowed, and than set if allowed, otherwise throw validation exception

Des135 commented on book-service-rest/src/main/java/com/example/Library/controller/UsersController.java:
```diff
@@ -0,0 +1,106 @@
+        if (gender.equals("FEMALE")) {
+            userDTO.setGender("FEMALE");
+        }
+        if (role.equals("ADMIN")) {
```
Same for roles validation

Des135 commented on book-service-rest/src/main/java/com/example/Library/controller/UsersController.java:
```diff
@@ -0,0 +1,106 @@
+    @PreAuthorize("hasAuthority('books:create')")
+    @PostMapping()
+    @Operation(summary = "Add a user to the library using parameters.")
+    public UserDTO addUser(@RequestParam(value = "firstName") String firstName,
```
wouldn't it be easier to accept UserDTO as an input, instead of many separate parameters? And then do a validation of a DTO

Des135 commented on book-service-rest/src/main/java/com/example/Library/exceptions/IncorrectGenderException.java:
```diff
@@ -0,0 +1,7 @@
+package com.example.Library.exceptions;
+
+public class IncorrectGenderException extends RuntimeException{
```
It seems like a common validation exception, many individual validation exceptions could be replaced with one common. (so only validation massage will be different)

Des135 commented on book-service-rest/src/main/java/com/example/Library/exceptions/IncorrectRoleException.java:
```diff
@@ -0,0 +1,8 @@
+package com.example.Library.exceptions;
+
+public class IncorrectRoleException extends RuntimeException{
```
same comment here, could be replaced by common validation exception

Des135 commented on book-service-rest/src/main/java/com/example/Library/exceptions/IncorrectUserStatusException.java:
```diff
@@ -0,0 +1,7 @@
+package com.example.Library.exceptions;
+
+public class IncorrectUserStatusException extends RuntimeException {
```
same comment here, could be replaced by common validation exception

Des135 commented on book-service-rest/src/main/java/com/example/Library/services/impl/LoginAttemptService.java:
```diff
@@ -0,0 +1,62 @@
+    public void loginSucceeded(String key) {
+        attemptsCache.invalidate(key);
+    }
+    public ArrayList<String> getAllBlockedUsers(){
```
please fix the spacing between the methods, it should be 1 new empty line between declared methods

Des135 commented on book-service-rest/src/main/java/com/example/Library/services/impl/LoginAttemptService.java:
```diff
@@ -0,0 +1,62 @@
+        }
+        return allBlockedUser;
+    }
+
```
same issue with code formatting here

Des135 commented on book-service-rest/src/main/java/com/example/Library/exceptions/UserAlreadyExists.java:
```diff
@@ -0,0 +1,7 @@
+package com.example.Library.exceptions;
+
+public class UserAlreadyExists extends RuntimeException {
```
do you have a proper exception handing for this Exception? It is a good practice to return 409 response code in case of such conflicts

Des135 commented on book-service-rest/src/main/java/com/example/Library/exceptions/UserIsBlocked.java:
```diff
@@ -0,0 +1,7 @@
+package com.example.Library.exceptions;
+
+public class UserIsBlocked extends RuntimeException {
```
it would be good to return 403 (Forbidden) response code in case of this Exception

Des135 commented on book-service-rest/src/main/java/com/example/Library/controller/UsersController.java:
```diff
@@ -0,0 +1,106 @@
+        return loginAttemptService.getAllBlockedUsers();
+    }
+
+    @PreAuthorize("hasAuthority('books:create')")
```
'books:create' seems to be not a proper authority for a user creation. Feel free to create a new one for this purpose

Commit:
-The absolute path is not used in the source code.
-UserController.class. was refactored.
-Exceptions were added to BookAndUserExceptionsHandler.class.
